ch1.adoc

```diff
@@ -0,0 +1,521 @@
+= Introduction
+
+This tutorial will introduce you to write your own programming
+language using Cranelift as a backend.
+A compiler is usually divided between a frontend and a backend.
+The frontend parses the source language, convert it to an abstract
+syntax tree (AST), do the semantic analysis and convert the AST to an
+intermediate representation.
+The backend takes this intermediate representation and produce
+optimized machine code.
+
+Cranelift is a machine code generator written in Rust and can be used
+to write a just-in-time (JIT) compiler as well as to generate object files.
+
+We'll first write a JIT compiler for the Kaleidoscope language.
+The syntax of this language and the Kaleidoscope code examples come from the
+https://llvm.org/docs/tutorial/index.html[LLVM tutorial].
+Then, we'll transform the JIT compile to an ahead-of-time compiler.
+The resulting compiler has around 1000 lines of code.
+
+== The Kaleidoscope language
+
+Kaleidoscope is a basic language with only one type: 64-bit floating point.
+Thus, it does not need type annotations or type inference which help
+to keep the language simple.
+Here's what it looks like:
+
+[source,python]
+----
+# Compute the x'th fibonacci number.
+def fib(x)
+  if x < 3 then
+    1
+  else
+    fib(x-1)+fib(x-2)
+
+# This expression will compute the 40th number.
+fib(40)
+----
+
+As you can see, it supports comments, functions, conditions and
+arithmetic.
+It can also call external functions as you can see in the following
+example:
+
+[source,python]
+----
+extern sin(arg);
+extern cos(arg);
+extern atan2(arg1 arg2);
+
+atan2(sin(.4), cos(42))
+----
+
+== Implementing the Lexer
```

Comment:
As a suggestion, it may be worth mentioning the option of using a parser-generator, such as lalrpop or peg, and that some of these don't require writing a lexer by hand. Kaleidoscope's syntax is so simple, we can do it by hand without too much trouble.